stan/math/opencl/kernel_generator/load.hpp

```diff
@@ -0,0 +1,156 @@
+#ifndef STAN_MATH_OPENCL_KERNEL_GENERATOR_LOAD_HPP
+#define STAN_MATH_OPENCL_KERNEL_GENERATOR_LOAD_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/matrix_cl_view.hpp>
+#include <stan/math/opencl/kernel_generator/type_str.hpp>
+#include <stan/math/opencl/kernel_generator/name_generator.hpp>
+#include <stan/math/opencl/kernel_generator/operation.hpp>
+#include <type_traits>
+#include <string>
+#include <utility>
+#include <set>
+
+namespace stan {
+namespace math {
+
+/**
+ * Represents an access to a \c matrix_cl in kernel generator expressions
+ * @tparam T \c matrix_cl
+ */
+template <typename T>
+class load__
+    : public operation<load__<T>, typename std::remove_reference_t<T>::type> {
+ public:
+  using ReturnScalar = typename std::remove_reference_t<T>::type;
+  using base = operation<load__<T>, ReturnScalar>;
+  using base::instance;
+  using base::var_name;
+  static_assert(std::is_base_of<matrix_cl<ReturnScalar>,
+                                typename std::remove_reference_t<T>>::value,
+                "load__: argument a must be a matrix_cl<T>!");
+  static_assert(
+      std::is_arithmetic<ReturnScalar>::value,
+      "load__: T in \"matrix_cl<T> a\" argument must be an arithmetic type!");
+
+  /**
+   * Constructor
+   * @param a \c matrix_cl
+   */
+  explicit load__(T&& a) : a_(std::forward<T>(a)) {}
+
+  /**
+   * generates kernel code for this expression.
+   * @param ng name generator for this kernel
+   * @param[in,out] generated set of already generated operations
+   * @param i row index variable name
+   * @param j column index variable name
+   * @return part of kernel with code for this and nested expressions
+   */
+  inline kernel_parts generate(name_generator& ng, std::set<int>& generated,
+                               const std::string& i,
+                               const std::string& j) const {
+    if (generated.count(instance) == 0) {
+      generated.insert(instance);
+      var_name = ng.generate();
+      kernel_parts res;
+      std::string type = type_str<ReturnScalar>::name;
+      res.body = type + " " + var_name + " = 0;"
+                 " if (!((!contains_nonzero(" + var_name + "_view, LOWER) && "
+                 + j + " < " + i + ") || (!contains_nonzero(" + var_name +
+                 "_view, UPPER) && " + j + " > " + i + "))) {"
+                 + var_name + " = " + var_name + "_global[" + i + " + " +
+                 var_name + "_rows * " + j + "];}\n";
+      res.args = "__global " + type + "* " + var_name + "_global, int "
+                 + var_name + "_rows, int " + var_name + "_view, ";
```

Comment:
Can you show me what the output from some of this code generation looks like?

Reply:
```
matrix_cl<double> a,b,c;
matrix_cl<double> d = a + b - 3 * c;
```
results in
```
kernel void calculate(__global double* var1_global, int var1_rows, int var1_view, __global double* var2_global, int var2
_rows, int var2_view, int var4, __global double* var5_global, int var5_rows, int var5_view, __global double* var8_global
, int var8_rows, int var8_view){
int i = get_global_id(0);int j = get_global_id(1);
double var1 = 0; if (!((!contains_nonzero(var1_view, LOWER) && j < i) || (!contains_nonzero(var1_view, UPPER) && j > i))
) {var1 = var1_global[i + var1_rows * j];}
double var2 = 0; if (!((!contains_nonzero(var2_view, LOWER) && j < i) || (!contains_nonzero(var2_view, UPPER) && j > i))
) {var2 = var2_global[i + var2_rows * j];}
double var3 = var1+var2;
double var5 = 0; if (!((!contains_nonzero(var5_view, LOWER) && j < i) || (!contains_nonzero(var5_view, UPPER) && j > i))
) {var5 = var5_global[i + var5_rows * j];}
double var6 = var4*var5;
double var7 = var3-var6;
var8_global[i + var8_rows * j] = var7;}
```